Content Cards Logging Analytics (#5545)
braze-inc/braze-docs: _docs/_developer_guide/customization_guides/content_cards/logging_analytics.md

article_title: Logging Analytics 
page_order: 4
description: "This article covers how to manually log clicks, events, and analytics for your customized Content Cards."
toc_headers: "h2"

---

# Logging analytics manually

> > If you would like to display the Content Cards in a completely custom manner, you can implement your own Content Cards presentation UI. However, data such as impressions, clicks, and dismissals are only handled automatically when using Braze's default card models. When implementing completely custom UI, you will need to handle this data manually. To do this, populate your custom UI with data from Braze's data models and manually log analytics like impressions and clicks. 

Reviewer: ```suggestion
> If you would like to display the Content Cards in a completely custom manner, you can implement your own Content Cards presentation UI. However, data such as impressions, clicks, and dismissals are only handled automatically when using the default card models. When implementing completely custom UI, you will need to handle this data manually. To do this, populate your custom UI with data from the data models and manually log analytics like impressions and clicks. 
```

## Listening for card updates

> When implementing your custom Content Cards, you can parse Braze's Content Card objects and extract their payload data such as `title`, `cardDescription`, and `imageUrl`. Then, you can use the resulting model data to populate your custom UI. 

Reviewer: ```suggestion
When implementing your custom Content Cards, you can parse the Content Card objects and extract their payload data such as `title`, `cardDescription`, and `imageUrl`. Then, you can use the resulting model data to populate your custom UI. 
```

> To obtain Braze's Content Card data model, subscribe to Content Card updates. There are two properties to pay particular attention to:

Reviewer: ```suggestion
To obtain the Content Card data models, subscribe to Content Card updates. There are two properties to pay particular attention to:
```

article_title: Logging Analytics 
page_order: 4
description: "This article covers how to manually log clicks, events, and analytics for your customized Content Cards."
toc_headers: "h2"

---

# Logging analytics manually

> If you would like to display the Content Cards in a completely custom manner, you can implement your own Content Cards presentation UI. However, data such as impressions, clicks, and dismissals are only handled automatically when using Braze's default card models. When implementing completely custom UI, you will need to handle this data manually. To do this, populate your custom UI with data from Braze's data models and manually log analytics like impressions and clicks. 

## Listening for card updates

When implementing your custom Content Cards, you can parse Braze's Content Card objects and extract their payload data such as `title`, `cardDescription`, and `imageUrl`. Then, you can use the resulting model data to populate your custom UI. 

To obtain Braze's Content Card data model, subscribe to Content Card updates. There are two properties to pay particular attention to:

> * **`idString`**: Represents the Content Card ID. This is the unique identifier that must always be parsed in order to log analytics from custom Content Cards.

Reviewer 2: ```suggestion
* **`idString`**: Represents the Content Card ID. This is the unique identifier used to log analytics from custom Content Cards.
```

* **`extras`**: Encompasses all the key-value pairs from the Braze dashboard.

> All properties outside of `idString` and `extras` are optional to parse for custom Content Cards. For more information on the data model, see each platform's integration article: [Android][1], [iOS][2], [Web][3].

Reviewer 3: Technically speaking, there's no field on a Content Card literally called `idString`. It's just `id`.

Author: Got it. Thanks for that, good catch.

> <!--This alert was taken from the Web guide. Is it applicable for Android and Swift as well? -->

Reviewer 3: aside: Instead of leaving these comments in the code, can you leave them as Github comments instead? It would make them much easier to track and not miss.

For Android this is not the case. CC will refresh regardless of any listeners being present or not.

Author: Got it. Valuable feedback. Will do that going forward.

BRZCancellable *cancellable = [self.braze.contentCards subscribeToUpdates:^(NSArray<BRZContentCardRaw *> *contentCards) {
  // Implement your completion handler to respond to updates in `contentCards`.
}];
```

{% endsubtab %}
{% endsubtabs %}
{% endtab %}
{% tab Web %}

Register a callback function to subscribe for updates when cards are refreshed.

```javascript
import * as braze from "@braze/web-sdk";

braze.subscribeToContentCardsUpdates(function(updates){
  const cards = updates.cards;
  // do something with the latest instance of `cards`
});

braze.openSession();
```

{% endtab %}
{% endtabs %}

## Logging events

After extending your custom objects to function as Content Cards, logging valuable metrics like impressions, clicks, and dismissals is quick and simple. Set a custom click listener to manually handle these analytics.

<!-- JOSH TO DO: Reminder: When you remove the information from the Android guide that is duplicative with this article, move the Helper file information from Advanced Integration Guide into the Integration section --->

{% tabs %}
{% tab Android %}

> <!-- Android Question: What is the difference between this prescribed method and the advice given in: https://www.braze.com/docs/developer_guide/platform_integration_guides/android/content_cards/customization/handling_clicks_manually. Would that article's content be better to use? --->

Reviewer 3: @josh-mccrowell-braze can you elaborate on this question? It seems like neither the "contentCardable" section nor the https://www.braze.com/docs/developer_guide/platform_integration_guides/android/content_cards/customization/handling_clicks_manually content would be applicable here. https://www.braze.com/docs/developer_guide/platform_integration_guides/android/content_cards/customization/handling_clicks_manually is just describing how to set a listener for when cards are clicked or dismissed. This section is describing how to click/dismiss.

Author: First off, thanks for indulging what are probably simple questions--but they help me learn, and I never want to make assumptions. 
The content I pulled into this article is originally from the Advanced Implementation Guide (https://www.braze.com/docs/developer_guide/platform_integration_guides/android/content_cards/implementation_guide/). The content in Handling Clicks Manually article (https://www.braze.com/docs/developer_guide/platform_integration_guides/android/content_cards/customization/handling_clicks_manually) also is about data and analytics. 

My question is two part:1) What content should we be using for this article in particular?2) What are the use cases of setting a custom click listener vs extending a custom object to call the content card logging methods?

Reviewer 3: 1. The rest of this article is fine. It's just that this specific section is on `logging events`, which imo is completely answered in the bit starting with `To log an impression or click on a card, call...`. The part before that about the sample app, contentcardable, etc. seem like overkill when you can just call `Card.logClick()`.
2. 
> `What are the use cases of setting a custom click listener `
I would say that the use-case here is something like `when this card is shown, do something` which you could either do with a full-blown custom integration, or by simply creating that listener and listening for clicks.

>vs extending a custom object to call the content card logging methods
I think that if you _only_ want to call the logging methods, then the custom object route is overkill. The logging methods are just `card.LogClick` so a custom wrapper is more work. If someone wants to have more control for whatever reason, then adding custom logging methods to their customized wrapper would make sense for their own convenience.

Author: Heard on it being overkill, I thought it looked odd but thought the Advanced Implementation Guide example might have been particularly thorough or expansive on the subject. I just pushed a revision that mostly leaned on the content from the Handling Clicks Manually article. Is that the content you would expect to see here? 

(The other content will continue to live in the Advanced Implementation article in the Android integration section.)

Reviewer 3: yup! The new changes lgtm.

Author: Thanks!
